src/libmeasurement_kit/ooni/meek_fronted_requests.cpp

```diff
@@ -0,0 +1,88 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "../common/utils.hpp"
+#include "../ooni/constants.hpp"
+#include <measurement_kit/ooni.hpp>
+
+namespace mk {
+namespace ooni {
+
+using namespace mk::report;
+
+void meek_fronted_requests(std::string input, Settings options,
+                   Callback<Var<report::Entry>> callback,
+                   Var<Reactor> reactor, Var<Logger> logger) {
+    Var<Entry> entry(new Entry);
+
+    std::string expected_body, outer_host, inner_host;
+
+    if (options["expected_body"].empty()) {
+        expected_body = constants::MEEK_SERVER_RESPONSE;
+    } else {
+        expected_body = options["expected_body"];
+    }
+
+    std::list<std::string> outer_inner = split(input, ":");
+    if (outer_inner.size() != 2) {
+        std::string fail_msg = "Couldn't split input: " + input;
+        logger->warn(fail_msg.c_str());
+        (*entry)["failure"] = fail_msg;
+        callback(entry);
+        return;
+    }
+    // XXX: We should make sure that we remove leading and trailing whitespaces
+    outer_host = outer_inner.front();
+    inner_host = outer_inner.back();
+
+    // url parsing methods require a schema
+    outer_host.insert(0, "https://");
+    inner_host.insert(0, "https://");
+
+    ErrorOr<http::Url> outer_url = mk::http::parse_url_noexcept(outer_host);
+    ErrorOr<http::Url> inner_url = mk::http::parse_url_noexcept(inner_host);
+
+    if (!outer_url || !inner_url) {
+        std::string fail_msg = "Invalid url: " + outer_host + " or "
+                                               + inner_host;
+        logger->warn(fail_msg.c_str());
+        (*entry)["failure"] = fail_msg;
```

Comment:
Regarding the error returned here, you can either use `ValueError` as above, or, as another option, you can perhaps extract the error from `inner_url` or `outer_url` using e.g. `inner_url.as_error()`.